MaxText/experimental/rl/grpo_trainer.py

```diff
@@ -0,0 +1,883 @@
+  group_mean = jnp.mean(rewards_grouped, axis=1)  # shape [B]
+  group_std = jnp.std(rewards_grouped, axis=1)  # shape [B]
+  repeated_group_mean = jnp.repeat(group_mean, G)  # shape [BxG]
+  repeated_group_std = jnp.repeat(group_std, G)  # shape [BxG]
+  advantages = (rewards - repeated_group_mean) / (repeated_group_std + 1e-4)  # shape [BxG]
+
+  # --- (8) Compute per-token loss.
+  # We follow the TRL GRPO loss:
+  #   loss_token = - [ exp(policy_logp - stop_gradient(policy_logp)) * advantage - beta * kl ]
+  # Make sure to expand advantage along the token dimension.
+  advantages_exp = advantages[:, None]  # shape [BxG, 1]
+
+  policy_diff = token_logps_policy - jax.lax.stop_gradient(token_logps_policy)
+  loss_tokens = -(jnp.exp(policy_diff) * advantages_exp - config.grpo_beta * per_token_kl)
+
+  # Average over tokens per generated completion.
+  loss_per_example = jnp.sum(loss_tokens * valid_seq_mask, axis=1) / (jnp.sum(valid_seq_mask, axis=1) + EPS)
+
+  loss = jnp.mean(loss_per_example)
+  total_weights = jnp.sum(valid_seq_mask)
+
+  moe_lb_loss = 0.0
+  if config.num_experts > 1:
+    nested_key = ("intermediates", "decoder", "layers", "moe_lb_loss")
+    total_moe_lb_loss = maxtext_utils.get_nested_value(intermediate_outputs, nested_key, 0.0)
+    moe_lb_loss = jnp.mean(jnp.array(total_moe_lb_loss))
+    loss += moe_lb_loss
+
+  # --- (9) Compute auxiliary metrics.
+  avg_kl = jnp.mean((per_token_kl * valid_seq_mask) / (jnp.sum(valid_seq_mask, axis=1, keepdims=True) + EPS))
+  avg_reward = jnp.mean(rewards)
+  avg_advantage = jnp.mean(advantages)
+  avg_completion_length = jnp.mean(jnp.sum(data["ar_completions_segmentation"] != 0, axis=1))
+  aux = {
+      "total_loss": loss,
+      "avg_reward": avg_reward,
+      "avg_reward_std": jnp.mean(repeated_group_std),
+      "avg_advantage": avg_advantage,
+      "avg_kl": avg_kl,
+      "completion_length": avg_completion_length,
+      "moe_lb_loss": moe_lb_loss,
+      "total_weights": total_weights,
+  }
+
+  return loss, aux
+
+
+# --- GRPO Helpers ---
+
+
+def generate_completions(config, tokenizer_model, engine, data, params, true_length, rng):
+  """
+  Autoregressively generates completions for a batch of prompts.
+
+  Args:
+    prompts: Array of shape [B, S] containing token ids.
+    config: Configuration containing:
+         - num_generations: number of completions to generate per prompt.
+         - max_completion_length: maximum number of tokens to generate.
+         - temperature: sampling temperature.
+    rng: JAX PRNGKeys.
+    tokenizer_model: Tokenizer for generate
+    true_length: Length of the prompt out of the max_target_length
+
+  Returns:
+    A jnp.array of shape [B x num_generations, S] where S = length_of_prompt + max_completion_length.
+  """
+  # decimate proportion of data when per_device_batch_size<1
+  for k, v in data.items():
+    if v.ndim == 2:
+      data[k] = v[: config.micro_batch_size_to_train_on, :]
```

Comment:
i believe it's more idiomatic to do :to_num. In fact reading it with current spacing I mis-read it as `:, bsz, :`